en: add the doc about upgrading from 1.0 to 2.0 manually (#213)
pingcap/docs-dm: en/manually-upgrade-dm-1.0-to-2.0.md

---

> title: Manually Upgrade TiDB Data Migration from 1.0.x to 2.0.x
> summary: Learn how to manually upgrade TiDB data migration from 1.0.x to 2.0.x.
> ---
> 
> # Manually Upgrade TiDB Data Migration from 1.0.x to 2.0.x

Reviewer: ```suggestion
title: Manually Upgrade TiDB Data Migration from v1.0.x to v2.0.x
summary: Learn how to manually upgrade TiDB data migration from v1.0.x to v2.0.x.
---

# Manually Upgrade TiDB Data Migration from v1.0.x to v2.0.x
```
To be consistent

> This document introduces how to manually upgrade TiDB migration task from v1.0.x to v2.0.x. The main idea is to use the global checkpoint information in v1.0.x to start a new data replication task in the v2.0.x cluster.

Reviewer: ```suggestion
This document introduces how to manually upgrade the TiDB DM tool from v1.0.x to v2.0.x. The main idea is to use the global checkpoint information in v1.0.x to start a new data replication task in the v2.0.x cluster.
```

> **Note:**
>

> > - Currently, DM does not support upgrading from v1.0.x to v2.0.x when the data migration task is in the process of full export or full import.

Reviewer: ```suggestion
> - Currently, upgrading DM from v1.0.x to v2.0.x is not supported when the data migration task is in the process of full export or full import.
```

> > - As the gRPC protocol used for interaction between the components of the DM cluster is renewed greatly, you need to make sure that the DM components (including dmctl) use the same version before and after the upgrade.

Reviewer: ```suggestion
> - As the gRPC protocol used for interaction between the components of the DM cluster is updated greatly, you need to make sure that the DM components (including dmctl) use the same version before and after the upgrade.
```

> > - Because the metadata storage of the DM cluster (such as checkpoint, shard DDL lock status and online DDL metadata, etc.) is renewed greatly, the metadata of v1.0.x cannot be reused automatically in v2.0.x. So you need to make sure the following requirements are satisfied before performing the upgrade operation:

Reviewer: ```suggestion
> - Because the metadata storage of the DM cluster (such as checkpoint, shard DDL lock status and online DDL metadata, etc.) is updated greatly, the metadata of v1.0.x cannot be reused automatically in v2.0.x. So you need to make sure the following requirements are satisfied before performing the upgrade operation:
```

>     - All data migration tasks are not in the process of shard DDL coordination.
>     - All data migration tasks are not in the process of online DDL coordination.

> The steps for manual upgrade are as follow.

Reviewer: ```suggestion
The steps for manual upgrade are as follows.
```

## Step 1: Prepare v2.0.x configuration file

The prepared configuration files of v2.0.x include the configuration files of the upstream database and the configuration files of the data migration task.

### Upstream database configuration file

> In v2.0.x, the [upstream database configuration file](source-configuration-file.md) is separated from the process configuration of the DM-worker, so you need to obtain the source configuration based on the [v1.0.x DM-worker configuration](https://docs.pingcap.com/zh/tidb-data-migration/stable/dm-worker-configuration-file).

Reviewer 2: please link to the English version of `https://docs.pingcap.com/zh/tidb-data-migration/stable/dm-worker-configuration-file`.

Author: Done

> **Note:**
>
> If `enable-gtid` in the source configuration is enabled during the upgrade from v1.0.x to v2.0.x, you need to parse the binlog or relay log file to obtain the GTID sets corresponding to the binlog position.

> #### Upgrade from v1.0.x deployed by DM-Ansible

Reviewer: ```suggestion
#### Upgrade a v1.0.x cluster deployed by DM-Ansible
```

> Assume that v1.0.x is deployed by DM-Ansible, and the following `dm_worker_servers` configuration is in the `inventory.ini`:

Reviewer: ```suggestion
Assume that the v1.0.x DM cluster is deployed by DM-Ansible, and the following `dm_worker_servers` configuration is in the `inventory.ini` file:
```

```ini
[dm_master_servers]
dm_worker1 ansible_host=172.16.10.72 server_id=101 source_id="mysql-replica-01" mysql_host=172.16.10.81 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
dm_worker2 ansible_host=172.16.10.73 server_id=102 source_id="mysql-replica-02" mysql_host=172.16.10.82 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
```

> Then you can convert, and obtain the following two source configuration files:

Reviewer: ```suggestion
Then you can convert it to the following two source configuration files:
```

source-id: "mysql-replica-01"                    # Corresponds to the original `source_id`.
from:
  host: "172.16.10.81"                           # Corresponds to the original `mysql_host`.
  port: 3306                                     # Corresponds to the original `mysql_port`.
  user: "root"                                   # Corresponds to the original `mysql_user`.
  password: "VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU="   # Corresponds to the original `mysql_password`.
```

```yaml
# The source configuration corresponding to the original dm_worker2. For example, it is named as source2.yaml.
server-id: 102                                   # Corresponds to the original `server_id`.
source-id: "mysql-replica-02"                    # Corresponds to the original `source_id`.
from:
  host: "172.16.10.82"                           # Corresponds to the original `mysql_host`.
  port: 3306                                     # Corresponds to the original `mysql_port`.
  user: "root"                                   # Corresponds to the original `mysql_user`.
  password: "VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU="   # Corresponds to the original `mysql_password`.
```

> #### Upgrade from v1.0.x deployed by Binary

Reviewer: ```suggestion
#### Upgrade a v1.0.x cluster deployed by binary
```

> Assume that v1.0.x is deployed by Binary, and the corresponding DM-worker configuration is as follow:

Reviewer: ```suggestion
Assume that the v1.0.x DM cluster is deployed by binary, and the corresponding DM-worker configuration is as follows:
```

```toml
log-level = "info"
log-file = "dm-worker.log"
worker-addr = ":8262"
server-id = 101
source-id = "mysql-replica-01"
flavor = "mysql"
[from]
host = "172.16.10.81"
user = "root"
password = "VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU="
port = 3306
```

> Then you can convert, and obtain the following source configuration file:

Reviewer: ```suggestion
Then you can convert it to the following source configuration file:
```

```yaml
server-id: 101                                   # Corresponds to the original `server-id`.
source-id: "mysql-replica-01"                    # Corresponds to the original `source-id`.
flavor: "mysql"                                  # Corresponds to the original `flavor`.
from:
  host: "172.16.10.81"                           # Corresponds to the original `from.host`.
  port: 3306                                     # Corresponds to the original `from.port`.
  user: "root"                                   # Corresponds to the original `from.user`.
  password: "VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU="   # Corresponds to the original `from.password`.
```

### Data migration task configuration file

For [data migration task configuration file](task-configuration-file.md), v2.0.x is basically compatible with v1.0.x. You can directly copy the configuration of v1.0.x.

## Step 2: Deploy the v2.0.x cluster

Use TiUP to deploy a new v2.0.x cluster according to the required number of nodes.

## Step 3：Stop the v1.0.x cluster

> If the original v1.0.x cluster is deployed by DM-Ansible, you need to [use DM-Ansible to stop the v1.0.x cluster](https://docs.pingcap.com/zh/tidb-data-migration/stable/cluster-operations#%E4%B8%8B%E7%BA%BF%E9%9B%86%E7%BE%A4).

Reviewer 2: please link to the English version of `https://docs.pingcap.com/zh/tidb-data-migration/stable/cluster-operations#%E4%B8%8B%E7%BA%BF%E9%9B%86%E7%BE%A4`.

Author: Done

> If the original v1.0.x cluster is deployed by Binary, you can stop the DM-worker and DM-master processes directly.

Reviewer: ```suggestion

If the original v1.0.x cluster is deployed by binary, you can stop the DM-worker and DM-master processes directly.
```

## Step 4: Upgrade data migration task

> 1. Use the [`operate-source`](manage-source.md#load-the-data-source-configurations) command to load the upstream database source configuration that is obtained in the step 1 of [prepare v2.0.x configuration file](#step-1-prepare-v20x-configuration-file) into the v2.0.x cluster.

Reviewer: ```suggestion
1. Use the [`operate-source`](manage-source.md#load-the-data-source-configurations) command to load the upstream database source configuration from [step 1](#step-1-prepare-v20x-configuration-file) into the v2.0.x cluster.
```

> 2. In the downstream of TiDB, obtain the corresponding global checkpoint information from the incremental checkpoint table of the v1.0.x data migration task.

Reviewer: ```suggestion
2. In the downstream TiDB cluster, obtain the corresponding global checkpoint information from the incremental checkpoint table of the v1.0.x data migration task.
```

>     - Assume that the v1.0.x data migration configuration does not specify `meta-schema` (or specify its value as the default `dm_meta`), and the corresponding task name is `task_v1`, the corresponding checkpoint information is in the ``` `dm_meta`.`task_v1_syncer_checkpoint` ``` table of the downstream TiDB 

Reviewer: ```suggestion
    - Assume that the v1.0.x data migration configuration does not specify `meta-schema` (or specify its value as the default `dm_meta`), and the corresponding task name is `task_v1`, the corresponding checkpoint information is in the ``` `dm_meta`.`task_v1_syncer_checkpoint` ``` table of the downstream TiDB.
```

    - Use the following SQL statements to obtain the global checkpoint information of all upstream database sources corresponding to the data migration task.

        ```sql
        > SELECT `id`, `binlog_name`, `binlog_pos` FROM `dm_meta`.`task_v1_syncer_checkpoint` WHERE `is_global`=1;
        +------------------+-------------------------+------------+
        | id               | binlog_name             | binlog_pos |
        +------------------+-------------------------+------------+
        | mysql-replica-01 | mysql-bin|000001.000123 | 15847      |
        | mysql-replica-02 | mysql-bin|000001.000456 | 10485      |
        +------------------+-------------------------+------------+
        ```

3. Update the v1.0.x data migration task configuration file to start a new v2.0.x data migration task.

>     - If the data migration task configuration file of v1.0.x is `task_v1.yaml`, copy it as `task_v2.yaml`.

Reviewer: ```suggestion
    - If the data migration task configuration file of v1.0.x is `task_v1.yaml`, copy it and rename it to `task_v2.yaml`.
```

    - Make the following changes to `task_v2.yaml`:

>         - Modify `name` to a non-existent name, such as `task_v2`.

Reviewer: ```suggestion
        - Modify `name` to a new name, such as `task_v2`.
```

        - Change `task-mode` to `incremental`.

>         - Set the starting point of incremental migration for each source according to the global checkpoint information obtained in step 2. For example:

Reviewer: ```suggestion
        - Set the starting point of incremental replication for each source according to the global checkpoint information obtained in step 2. For example:
```
